### PR TITLE
dice loss

### PR DESCRIPTION
**Pull request by aashokvardhan**

### CODE REVIEW

File: fastestimator/backend/dice_loss.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        channel_average: Whether to average the channel wise loss.
+
+    Returns:
+        The dice loss between `y_pred` and `y_true`. A scalar if `average_sample_loss` is True, else a
+        tensor with the shape (Batch).
+
+    Raises:
+        AssertionError: If `y_true` or `y_pred` are unacceptable data types.
+    """
+
```
Comment (vbvg2008) [inline fastestimator/backend/dice_loss.py:162]: cast y_true to be the same type of y_pred

File: fastestimator/op/tensorop/loss/dice_loss.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                 channel_average: bool = False):
+        super().__init__(inputs=inputs, outputs=outputs, mode=mode,
+                         ds_id=ds_id, average_loss=average_sample_loss)
+        self.channel_average = channel_average
+        self.log_loss = log_loss
+        self.soft_dice_loss = soft_dice_loss
+
+    def forward(self, data: List[Tensor], state: Dict[str, Any]) -> Tensor:
+        y_pred, y_true = data
+        return dice_loss(y_pred, y_true, self.log_loss, self.soft_dice_loss, self.average_loss, self.channel_average)
```
Comment (vbvg2008) [inline fastestimator/op/tensorop/loss/dice_loss.py:64]: apply the 1-dice or log(dice) in this tensorOp

File: fastestimator/trace/metric/dice.py [hunk 3]
```diff
@@ -68,17 +73,14 @@ def on_epoch_begin(self, data: Data) -> None:
         self.dice = []
 
     def on_batch_end(self, data: Data) -> None:
-        y_true, y_pred = to_number(data[self.true_key]), to_number(data[self.pred_key])
-        batch_size = y_true.shape[0]
-        y_true, y_pred = y_true.reshape((batch_size, -1)), y_pred.reshape((batch_size, -1))
+        y_true, y_pred = data[self.true_key], data[self.pred_key]
```
Comment (vbvg2008) [inline fastestimator/trace/metric/dice.py:77]: apply the binary thresholding in the trace

**Review by vbvg2008 — CHANGES_REQUESTED** [review]

File: fastestimator/backend/dice_loss.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
+# Copyright 2022 The FastEstimator Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (aashokvardhan) [reply fastestimator/backend/dice_loss.py]: The file only calculates dice score.

File: fastestimator/backend/dice_loss.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+    """
+    dice = clip_by_value(dice, eps)
+    if log_loss:
+        if tf.is_tensor(dice):
+            return -tf.math.log(dice)
+        elif isinstance(dice, torch.Tensor):
+            return -torch.log(dice)
+        elif isinstance(dice, np.ndarray):
+            return -np.log(dice)
```
[the remark this answers is not in the document]
Comment (aashokvardhan) [reply fastestimator/backend/dice_loss.py]: MOved the logic to dice score.

File: fastestimator/backend/dice_score.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        channel_average: Whether to average the channel wise dice score.
+
+    Returns:
+        The dice score between `y_pred` and `y_true`. A scalar if `average_sample` is True, else a
+        tensor with the shape (Batch).
+
+    Raises:
+        AssertionError: If `y_true` or `y_pred` are unacceptable data types. if data type is other than np.array, tensor.Tensor, tf.Tensor.
+    """
+    y_true = convert_input_precision(y_true)
```
Comment (vbvg2008) [inline fastestimator/backend/dice_score.py:128]: converting precision here isn't necessary. 

cast y_true and epsilon to be the same type as y_red

File: fastestimator/backend/__init__.py [hunk 7]
```diff
@@ -20,6 +20,8 @@
 from fastestimator.backend.check_nan import check_nan
 from fastestimator.backend.clip_by_value import clip_by_value
 from fastestimator.backend.concat import concat
+from fastestimator.backend.convert_tensor_precision import convert_input_precision
```
Comment (vbvg2008) [inline fastestimator/backend/__init__.py:23]: avoid renaming import function

File: fastestimator/backend/dice_score.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            y_pred: Prediction with a shape like (Batch, C). dtype: float32 or float16.
+            y_true: Ground truth class labels with a shape like `y_pred`. dtype: int or float32 or float16.
+            soft_dice: Whether to add direct sum or square sum of inputs
+
+        Return:
+            The sum or squared sum of y_pred and y_true
+    """
+    if soft_dice:
+        if tf.is_tensor(y_true):
+            return tf.square(y_true) + tf.square(y_pred)
```
Comment (vbvg2008) [inline fastestimator/backend/dice_score.py:42]: **2

File: fastestimator/backend/dice_score.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    numerator = reduce_sum(y_true*y_pred, axis=axis)
+
+    denominator = get_denominator(y_true, y_pred, soft_dice)
+
+    denominator = reduce_sum(denominator, axis=axis)
+
+    dice_score = (2 * (numerator + epsilon)) / (denominator + epsilon)
+
+    if channel_average:
+        dice_score = reduce_mean(dice_score, axis=1)
```
Comment (vbvg2008) [inline fastestimator/backend/dice_score.py:143]: need to extend the dice to Nd, by assuming tf:[B, ..., C], torch:[B,C, ...]

File: fastestimator/op/tensorop/dice.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+Tensor = TypeVar('Tensor', tf.Tensor, torch.Tensor, np.array)
+
+
+class DiceScore(TensorOp):
+    """
+    Calculate Element-Wise Dice Score.
+
+        Args:
+            y_pred: Prediction with a shape like (Batch, C). dtype: float32 or float16.
```
Comment (vbvg2008) [inline fastestimator/op/tensorop/dice.py:32]: Docstring

File: fastestimator/trace/metric/dice.py [hunk 11]
```diff
@@ -68,15 +74,13 @@ def on_epoch_begin(self, data: Data) -> None:
...
+
+        y_pred = convert_input_precision(y_pred > self.threshold)
```
Comment (vbvg2008) [inline fastestimator/trace/metric/dice.py:79]: conversion is unnecessary

**Review by vbvg2008 — CHANGES_REQUESTED** [review]

File: fastestimator/op/tensorop/dice.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+import tensorflow as tf
+import torch
+
+from fastestimator.backend.dice_score import dice_score
+from fastestimator.op.tensorop.tensorop import TensorOp
+
+Tensor = TypeVar('Tensor', tf.Tensor, torch.Tensor, np.array)
+
+
+class DiceScore(TensorOp):
```
Comment (vbvg2008) [inline fastestimator/op/tensorop/dice.py:27]: let's call it Dice

File: fastestimator/op/tensorop/dice.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        self.channel_average = channel_average
+        self.soft_dice = soft_dice
+        self.epsilon = epsilon
+        self.sample_average = sample_average
+
+    def forward(self, data: List[Tensor], state: Dict[str, Any]) -> Tensor:
+        y_pred, y_true = data
+        dice = dice_score(
+            y_pred, y_true, self.soft_dice, self.sample_average, self.channel_average, self.epsilon)
+        return -dice
```
Comment (vbvg2008) [inline fastestimator/op/tensorop/dice.py:65]: let's have an argument `negate` with default False. 

When negate is False, we will return dice, when True, we will return -dice

**Review by vbvg2008 — CHANGES_REQUESTED** [review]
just 2 minor comments

**Review by vbvg2008 — APPROVED** [review]
LG

**Pull request merged**